File: hello-world-java/src/main/java/example/HelloFilter.java
package example;

import io.micronaut.http.HttpRequest;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Filter;
import io.micronaut.http.filter.HttpServerFilter;
import io.micronaut.http.filter.ServerFilterChain;
import org.reactivestreams.Publisher;

@Filter("/**")
public class HelloFilter implements HttpServerFilter {

    @Override
    public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request, ServerFilterChain chain) {

        System.out.println(String.format("%s [%s] doFilter(1) before proceed", Thread.currentThread().getName(), "HelloFilter"));

        Publisher<MutableHttpResponse<?>> responsePublisher = chain.proceed(request);

        System.out.println(String.format("%s [%s] doFilter(2) after proceed", Thread.currentThread().getName(), "HelloFilter"));

        responsePublisher.subscribe(new HelloResponseFilterSubscriber());

        System.out.println(String.format("%s [%s] doFilter(3) after subscribe", Thread.currentThread().getName(), "HelloFilter"));

        return responsePublisher;
    }
}

File: hello-world-kotlin/src/main/kotlin/example/HelloFilter.kt
package example

import io.micronaut.http.HttpRequest
import io.micronaut.http.MutableHttpResponse
import io.micronaut.http.annotation.Filter
import io.micronaut.http.filter.HttpServerFilter
import io.micronaut.http.filter.ServerFilterChain
import org.reactivestreams.Publisher

@Filter("/**")
class HelloFilter : HttpServerFilter {

    override fun doFilter(request: HttpRequest<*>, chain: ServerFilterChain): Publisher<MutableHttpResponse<*>> {

        println("${Thread.currentThread().name} [HelloFilter] doFilter (1) before proceed")

        val responsePublisher = chain.proceed(request)

        println("${Thread.currentThread().name} [HelloFilter] doFilter (2) after proceed")

        responsePublisher.subscribe(HelloResponseFilterSubscriber())

        println("${Thread.currentThread().name} [HelloFilter] doFilter (3) after subscribe")

        return responsePublisher
    }
}